### PR TITLE
Moved tournament api from v1 to v3

### PR DESCRIPTION
**Pull request by JanOuborny**
Appended `V1` to old methods.
Replaced `Tournament` and `TournamentProvider` with `int` in new methods.

### CODE REVIEW

**Comment by BenFradet** [comment]
Great stuff, I'll have a look tomorrow!

File: RiotSharp/ITournamentRiotApi.cs [hunk 1]
```diff
@@ -219,8 +373,10 @@ public interface ITournamentRiotApi
         /// <param name="pickType">The pick type.</param>
         /// <param name="mapType">The map type.</param>
         /// /// <returns>Success value.</returns>
-        Task<bool> UpdateTournamentCodeAsync(string tournamentCode, List<long> allowedSummonerIds,
+        [Obsolete]
+        Task<bool> UpdateTournamentCodeV1Async(string tournamentCode, List<long> allowedSummonerIds,
             TournamentSpectatorType? spectatorType, TournamentPickType? pickType, TournamentMapType? mapType);
+
+        #endregion
```
Comment (BenFradet) [inline RiotSharp/ITournamentRiotApi.cs]: Given the timeframe for porting everything to v3, I'd be in favor of making a clean cut removing those `V1` mehods that'll keep the codebase slimmer.

File: RiotSharp/TournamentEndpoint/Tournament.cs [hunk 2]
```diff
@@ -82,11 +87,12 @@ internal Tournament()
...
+        [Obsolete]
+        public async Task<List<string>> CreateTournamentCodesV1Async(int teamSize, TournamentSpectatorType spectatorType,
             TournamentPickType pickType, TournamentMapType mapType, string metadata, int count = 1)
         {
             return await TournamentRiotApi.GetInstance()
-                .CreateTournamentCodesAsync(Id, teamSize, spectatorType, pickType, mapType, metadata, count);
+                .CreateTournamentCodesV1Async(Id, teamSize, spectatorType, pickType, mapType, metadata, count);
         }
     }
 }
```
Comment (BenFradet) [inline RiotSharp/TournamentEndpoint/Tournament.cs]: same here I think removing the entire class is the way to go

File: RiotSharp/TournamentEndpoint/TournamentCodeDetail.cs [hunk 3]
```diff
@@ -114,31 +115,34 @@ internal TournamentCodeDetail()
...
         ///     Returns the tournament code object associated with a tournament code string asynchronously.
         /// </summary>
         /// <param name="tournamentCode">The tournament code.</param>
         /// <returns>the tournament code object associated with a tournament code string.</returns>
-        public static async Task<TournamentCodeDetail> GetAsync(string tournamentCode)
+        [Obsolete]
+        public static async Task<TournamentCodeDetail> GetV1Async(string tournamentCode)
         {
-            return await TournamentRiotApi.GetInstance().GetTournamentCodeDetailsAsync(tournamentCode);
+            return await TournamentRiotApi.GetInstance().GetTournamentCodeDetailsV1Async(tournamentCode);
```
Comment (BenFradet) [inline RiotSharp/TournamentEndpoint/TournamentCodeDetail.cs]: same here let's remove those.

File: RiotSharp/TournamentEndpoint/TournamentProvider.cs [hunk 4]
```diff
@@ -22,19 +24,21 @@ internal TournamentProvider()
...
         /// <returns>A tournament instance.</returns>
-        public async Task<Tournament> CreateTournamentAsync(string name)
+        [Obsolete]
+        public async Task<Tournament> CreateTournamentV1Async(string name)
         {
-            return await TournamentRiotApi.GetInstance().CreateTournamentAsync(Id, name);
+            return await TournamentRiotApi.GetInstance().CreateTournamentV1Async(Id, name);
         }
     }
 }
```
Comment (BenFradet) [inline RiotSharp/TournamentEndpoint/TournamentProvider.cs]: same, let's remove the entire class

File: RiotSharp/TournamentRiotApi.cs [hunk 5]
```diff
@@ -13,13 +13,20 @@ public class TournamentRiotApi : ITournamentRiotApi
...
+        private const string GetCodeUrlV1 = "/code/{0}";
+        private const string GetCodesUrl = "/codes/{0}";
+        private const string PutCodeUrlV1 = "/code/{0}";
+        private const string PutCodeUrl = "/codes/{0}";
+        private const string LobbyEventUrlV1 = "/lobby/events/by-code/{0}";
+        private const string LobbyEventUrl = "/lol/tournament/v3/lobby-events/by-code/{0}";
+        private const string CreateProviderUrlV1 = "/provider";
+        private const string CreateProviderUrl = "/providers";
+        private const string CreateTournamentUrlV1 = "/tournament";
+        private const string CreateTournamentUrl = "/tournaments";
```
Comment (BenFradet) [inline RiotSharp/TournamentRiotApi.cs]: same here let's remove all those *V1 urls

File: RiotSharp/TournamentRiotApi.cs [hunk 6]
```diff
@@ -69,58 +76,311 @@ internal static TournamentRiotApi GetInstance()
...
+        {
+            var body = new Dictionary<string, object>
+            {
+                { "name", name },
+                { "providerId", providerId }
+            };
+            var json = requester.CreatePostRequest(TournamentRootUrl + CreateTournamentUrl, Region.global,
+                JsonConvert.SerializeObject(body));
+
+            // Json is an int directly
```
Comment (BenFradet) [inline RiotSharp/TournamentRiotApi.cs]: no need for this comment id say.

File: RiotSharp/TournamentRiotApi.cs [hunk 7]
```diff
@@ -69,58 +76,311 @@ internal static TournamentRiotApi GetInstance()
...
+            ValidateTeamSize(teamSize);
+            ValidateTournamentCodeCount(count);
+            var body = new Dictionary<string, object>
+            {
+                {"teamSize", teamSize},
+                {"allowedSummonerIds", allowedParticipantIds},
+                {"spectatorType", spectatorType},
+                {"pickType", pickType},
+                {"mapType", mapType},
+                {"metadata", metadata}
```
Comment (BenFradet) [inline RiotSharp/TournamentRiotApi.cs]: can we keep things consistent and do:

```c#
{ "teamSize", teamSize },
{ "allowedSummonerIds", allowedParticipantIds },
{ "spectatorType", spectatorType },
{ "pickType", pickType },
{ "mapType", mapType },
{ "metadata", metadata }
```

File: RiotSharp/TournamentRiotApi.cs [hunk 8]
```diff
@@ -69,58 +76,311 @@ internal static TournamentRiotApi GetInstance()
...
+            };
+            var json = requester.CreatePostRequest(TournamentRootUrl + CreateCodesUrl, Region.global,
+                JsonConvert.SerializeObject(body, null, new JsonSerializerSettings { NullValueHandling = NullValueHandling.Ignore }),
+                new List<string> { string.Format("tournamentId={0}", tournamentId), string.Format("count={0}", count) });
+
+            // Json is a list of strings
```
Comment (BenFradet) [inline RiotSharp/TournamentRiotApi.cs]: no need for this comment I feel

File: RiotSharp/TournamentRiotApi.cs [hunk 9]
```diff
@@ -359,5 +581,8 @@ private void ValidateTournamentCodeCount(int count)
             if (count < 1)
                 throw new ArgumentException("Invalid count. The value of count must be greater than 0.", nameof(count));
         }
+
+        #endregion
+
```
Comment (BenFradet) [inline RiotSharp/TournamentRiotApi.cs]: here as well I think we're better off removing v1-related stuff

File: RiotSharp/ITournamentRiotApi.cs [hunk 10]
```diff
@@ -11,6 +12,182 @@ namespace RiotSharp
...
+        ///     Creates a tournament provider and returns its ID.
+        /// </summary>
+        /// <param name="region">The region in which the provider will be running tournaments.</param>
+        /// <param name="url">
+        ///     The provider's callback URL to which tournament game results in this region should be posted. The URL
+        ///     must be well-formed, use the http or https protocol, and use the default port for the protocol (http URLs must use
+        ///     port 80, https URLs must use port 443).
+        /// </param>
+        /// <returns>The ID of the provider.</returns>
+        int CreateProvider(Region region, string url);
```
Comment (BenFradet) [inline RiotSharp/ITournamentRiotApi.cs]: this nicely simplify things 👍

File: RiotSharpTest/TournamentRiotApiTest.cs [hunk 11]
```diff
@@ -260,8 +250,8 @@ public void CreateTournamentCodeAsync_InvalidTeamSize_ThrowsArgumentException()
             try
             {
                 // Act 
-                var tournamentCodes = api.CreateTournamentCodeAsync(0, 0, TournamentSpectatorType.All, 
-                    TournamentPickType.TournamentDraft, TournamentMapType.SummonersRift).Result;
+                var tournamentCodes = api.CreateTournamentCodeV1Async(0, 0, null, TournamentSpectatorType.All,
```
Comment (BenFradet) [inline RiotSharpTest/TournamentRiotApiTest.cs]: same thing here let's remove v1

File: RiotSharp/ITournamentRiotApi.cs [hunk 12]
```diff
@@ -3,6 +3,7 @@
 using RiotSharp.MatchEndpoint;
 using RiotSharp.TournamentEndpoint;
 using RiotSharp.TournamentEndpoint.Enums;
+using System;
```
Comment (BenFradet) [inline RiotSharp/ITournamentRiotApi.cs]: do we need this?

File: RiotSharp/ITournamentRiotApi.cs [hunk 13]
```diff
@@ -11,6 +12,8 @@ namespace RiotSharp
     /// </summary>
     public interface ITournamentRiotApi
     {
+        #region (Current) Version 3
+
```
Comment (BenFradet) [inline RiotSharp/ITournamentRiotApi.cs]: Since we've removed the v1 api I think we can do without

File: RiotSharp/TournamentRiotApi.cs [hunk 14]
```diff
@@ -69,157 +69,117 @@ internal static TournamentRiotApi GetInstance()
             }
             return instance;
         }
-    
-        public TournamentProvider CreateProvider(Region region, string url)
+
+        #region (Current) Version 3
```
Comment (BenFradet) [inline RiotSharp/TournamentRiotApi.cs]: same

File: RiotSharp/TournamentRiotApi.cs [hunk 15]
```diff
@@ -69,157 +69,117 @@ internal static TournamentRiotApi GetInstance()
...
+                { "allowedSummonerIds", allowedParticipantIds },
+                { "spectatorType", spectatorType },
+                { "pickType", pickType },
+                { "mapType", mapType },
+                { "metadata", metadata }
             };
-            var json = requester.CreatePostRequest(TournamentRootUrl + CreateCodeUrl, Region.global,
-                JsonConvert.SerializeObject(body),
+            var json = requester.CreatePostRequest(TournamentRootUrl + CreateCodesUrl, Region.global,
+                JsonConvert.SerializeObject(body, null, new JsonSerializerSettings { NullValueHandling = NullValueHandling.Ignore }),
```
Comment (BenFradet) [inline RiotSharp/TournamentRiotApi.cs]: could we reformat that so that it's easier to read?

File: RiotSharp/ITournamentRiotApi.cs (hunk 13 above)
Comment (JanOuborny) [inline RiotSharp/ITournamentRiotApi.cs]: But I like to minimize things :(

File: RiotSharp/TournamentRiotApi.cs (hunk 14 above)
Comment (JanOuborny) [inline RiotSharp/TournamentRiotApi.cs]: I think we should keep that region. It makes the code more readable. I'm just gonna rename it to `Public Methods`

File: RiotSharpTest/TournamentRiotApiTest.cs [hunk 16]
```diff
@@ -74,26 +71,19 @@ public void GetTournamentMatch_Test()
         [TestCategory("TournamentRiotApi"), TestCategory("Async")]
         public void CreateProviderAsync_CreateTournamentAsync_CreateTournamentCodeAsync_UpdateTournamentCodeAsync_Test()
```
Comment (BenFradet) [inline RiotSharpTest/TournamentRiotApiTest.cs]: I have this test failing, any idea why?

Comment (JanOuborny) [inline RiotSharpTest/TournamentRiotApiTest.cs]: This test fails if you run it simultaneously with the others. However, it passes if you run it solo. I don't know why, though.

Comment (BenFradet) [inline RiotSharpTest/TournamentRiotApiTest.cs]: Very weird indeed, could we ignore it and mention that it's not failing if ran alone?

Comment (BenFradet) [inline RiotSharpTest/TournamentRiotApiTest.cs]: Also could you rebase/merge so I can merge?

Comment (JanOuborny) [inline RiotSharpTest/TournamentRiotApiTest.cs]: Rebase is done and now I know why the test fails. It fails because of an exceeded rate limit,.

Comment (BenFradet) [inline RiotSharpTest/TournamentRiotApiTest.cs]: the rate limit isn't respected?

Comment (JanOuborny) [inline RiotSharpTest/TournamentRiotApiTest.cs]: @BenFradet  See my updated answer

**Pull request merged**